src/algorithms/PANOC.jl

```diff
@@ -0,0 +1,218 @@
+        Asxeps = sol.As*xeps
+        gradfs_Asxeps, = gradient(sol.fs, Asxeps)
+        At_gradf_Axeps = sol.As'*gradfs_Asxeps .+ sol.Aq'*gradfq_Aqxeps
+        L = blockvecnorm(sol.At_gradf_Ax .- At_gradf_Axeps)/(sqrt(eps()*blocklength(xeps)))
+        sol.adaptive = true
+        # in both cases set gamma = 1/L
+        sol.gamma = sol.alpha/L
+    end
+
+    blockaxpy!(sol.y, sol.x, -sol.gamma, sol.At_gradf_Ax)
+    sol.g_xbar = prox!(sol.xbar, sol.g, sol.y, sol.gamma)
+    blockaxpy!(sol.FPR_x, sol.x, -1.0, sol.xbar)
+
+    sol.normFPR_x = blockvecnorm(sol.FPR_x)
+    sol.FBE_x = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, sol.FPR_x) + 0.5/sol.gamma*sol.normFPR_x^2 + sol.g_xbar
+
+
+
+end
+
+################################################################################
+# Iteration
+
+function iterate!(sol::PANOCIterator{I, R, T}, it::I) where {I, R, T}
+
+	if sol.adaptive
+		for it_gam = 1:100 # TODO: replace/complement with lower bound on gamma
+			Aqxbar = sol.Aq*sol.xbar
+			gradfq_Aqxbar, fq_Aqxbar = gradient(sol.fq, Aqxbar)
+			Asxbar = sol.As*sol.xbar
+			gradfs_Asxbar, fs_Asxbar = gradient(sol.fs, Asxbar)
+			f_Axbar = fs_Asxbar + fq_Aqxbar
+
+			uppbnd = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, sol.FPR_x) + 
+				 0.5/sol.gamma*sol.normFPR_x^2
+			if f_Axbar > uppbnd + 1e-6*abs(sol.f_Ax)
+				sol.gamma = 0.5*sol.gamma
+				blockaxpy!(sol.y, sol.x, -sol.gamma, sol.At_gradf_Ax)
+				sol.g_xbar = prox!(sol.xbar, sol.g, sol.y, sol.gamma)
+				blockaxpy!(sol.FPR_x, sol.x, -1.0, sol.xbar)
+				sol.normFPR_x = blockvecnorm(sol.FPR_x)
+			else
+				sol.FBE_x = uppbnd + sol.g_xbar
+				break
+			end
+		end
+	end
+
+	if it > 1
+		update!(sol.H, sol.x, sol.x_prev, sol.FPR_x, sol.FPR_x_prev)
+	end
+	A_mul_B!(sol.d, sol.H, 0.0 .- sol.FPR_x) # TODO: not nice
+
+	sol.FPR_x_prev, sol.FPR_x = sol.FPR_x, sol.FPR_x_prev
+	blockset!(sol.x_prev, sol.x)
+
+	C = sol.sigma*sol.gamma*(1.0-sol.alpha)
+	maxit_tau = 10
+	tau = 1.0
+    
+	for it_tau = 1:maxit_tau # TODO: replace/complement with lower bound on tau
+
+		xnew = (1-tau).*sol.xbar .+ tau .* (sol.x.+sol.d)
+	   
+		# calculate new FBE in xnew
+
+		A_mul_B!(sol.Aqx, sol.Aq, xnew)
+		sol.fq_Aqx = gradient!(sol.gradfq_Aqx, sol.fq, sol.Aqx)
+		A_mul_B!(sol.Asx, sol.As, xnew)
+		sol.fs_Asx = gradient!(sol.gradfs_Asx, sol.fs, sol.Asx)
+		blockaxpy!(sol.At_gradf_Ax, sol.As'*sol.gradfs_Asx, 1.0, sol.Aq'*sol.gradfq_Aqx)
+		sol.f_Ax = sol.fs_Asx + sol.fq_Aqx
+
+		# gradient step
+		blockaxpy!(sol.y, xnew, -sol.gamma, sol.At_gradf_Ax)
+		# prox step
+		xnewbar, g_xnewbar = prox(sol.g, sol.y, sol.gamma)
+	    
+		FPR_xnew = xnew .- xnewbar
+		norm_FPRxnew = blockvecnorm(FPR_xnew)
+	    
+		FBE_xnew = sol.f_Ax - blockvecdot(sol.At_gradf_Ax, FPR_xnew) + 
+		           0.5/sol.gamma*norm_FPRxnew^2 + g_xnewbar
+
+		if FBE_xnew <= sol.FBE_x - (C/2)*sol.normFPR_x^2 || it_tau == maxit_tau
+			sol.normFPR_x = norm_FPRxnew
+			sol.FPR_x = FPR_xnew
+			sol.FBE_x = FBE_xnew
+			sol.x, sol.xbar = xnew, xnewbar
```

Comment:
And here careful: you must swap all these vector references (which will all be in sol).

Also: why not simply `break` here and do all these update operations outside after the loop? It looks more elegant and you can probably avoid checking `it_tau == maxit_tau` in the `if`.